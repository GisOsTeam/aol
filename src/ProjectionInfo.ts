import Projection from 'ol/proj/Projection';
import { register } from 'ol/proj/proj4';
import proj4 from 'proj4';
import { transformExtent, get as getProjection } from 'ol/proj';
import { Extent } from 'ol/extent';

const projMap = new Map<string, ProjectionInfo>();

export function getProjectionInfo(code: string): ProjectionInfo {
  return projMap.get(code);
}

export function getProjectionInfos(): ProjectionInfo[] {
  const projectionInfos: ProjectionInfo[] = [];
  projMap.forEach(projectionInfo => {
    projectionInfos.push(projectionInfo);
  });
  return projectionInfos;
}

export function addProjection(
  code: string,
  wkt?: string,
  lonLatValidity?: Extent,
  name?: string,
  remarks?: string
): ProjectionInfo {
  const projectionInfo = new ProjectionInfo();
  projectionInfo.code = code;
  projectionInfo.wkt = wkt;
  projectionInfo.lonLatValidity = lonLatValidity;
  projectionInfo.name = name;
  projectionInfo.remarks = remarks;
  proj4.defs(projectionInfo.code, projectionInfo.wkt);
  console.info('Register projection ' + projectionInfo.code + ' - ' + projectionInfo.name);
  register(proj4);
  projectionInfo.projection = getProjection(projectionInfo.code);
  if (Array.isArray(projectionInfo.lonLatValidity)) {
    const extent = transformExtent(projectionInfo.lonLatValidity, 'EPSG:4326', projectionInfo.projection);
    projectionInfo.projection.setExtent(extent);
  }
  projMap.set(projectionInfo.code, projectionInfo);
  return projectionInfo;
}

export class ProjectionInfo {
  public code: string;
  public wkt: string;
  public lonLatValidity: Extent;
  public name: string;
  public remarks: string;
  public projection: Projection;
}
