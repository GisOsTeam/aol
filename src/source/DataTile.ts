import { GeoTIFF } from './GeoTIFF';

export type DataTile = GeoTIFF;
