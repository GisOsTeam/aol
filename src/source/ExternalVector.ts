import { Vector, IVectorOptions } from './Vector';
import { SourceType, SourceTypeEnum } from './types/sourceType';
import { LayerType, LayerTypeEnum } from './types/layerType';

export interface IExternalVectorOptions extends IVectorOptions {}

export class ExternalVector extends Vector {
  /**
   *
   * @param param0, default useSpatialIndex is false for performance issue
   */
  constructor({ useSpatialIndex = false, ...options }: IExternalVectorOptions) {
    super({ ...options, useSpatialIndex });
    this.options = options;
    if (this.options.snapshotable != false) {
      this.options.snapshotable = true;
    }
    if (this.options.listable != false) {
      this.options.listable = true;
    }
    if (this.options.removable != false) {
      this.options.removable = true;
    }
  }

  public getSourceType(): SourceType {
    return SourceTypeEnum.ExternalVector;
  }

  public getSourceOptions(): IExternalVectorOptions {
    return this.options;
  }

  public setSourceOptions(options: IExternalVectorOptions): void {
    this.options = { ...options };
  }

  public getLayerType(): LayerType {
    return LayerTypeEnum.Vector;
  }

  public isSnapshotable(): boolean {
    return this.options.snapshotable;
  }

  public isListable(): boolean {
    return this.options.listable;
  }

  public isRemovable(): boolean {
    return this.options.removable;
  }
}
