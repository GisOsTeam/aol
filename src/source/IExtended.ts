import OlMap from 'ol/Map';
import Source, { Options } from 'ol/source/Source';
import Feature from 'ol/Feature';
import Geometry from 'ol/geom/Geometry';
import Projection from 'ol/proj/Projection';
import { SourceType } from './types/sourceType';
import { LayerType } from './types/layerType';
import { AndPre, IPredicate, Or } from '../filter/predicate';
import { FilterBuilder } from '../filter/FilterBuilder';

export interface ISnapshotOptions extends Options {
  snapshotable?: boolean;
  listable?: boolean;
  removable?: boolean;
}

export interface ISnapshotSource extends Source {
  getSourceType(): SourceType;

  getSourceOptions(): ISnapshotOptions;

  setSourceOptions(options: ISnapshotOptions): void;

  getLayerType(): LayerType;

  isSnapshotable(): boolean; // Can save snapshot
  isListable(): boolean; // Visible by others tools
  isRemovable(): boolean; // Removable from map by others tools
}

export interface IInitSource extends ISnapshotSource {
  init(): Promise<void>;
}

export interface IQuerySource extends ISnapshotSource {
  query(identifyRequest: IGisRequest, onlyVisible?: boolean): Promise<IQueryResponse>;

  retrieveFeature(id: number | string, featureProjection: Projection): Promise<Feature>;
}

export interface ILayerLegend {
  /**
   * Label optinal
   */
  label?: string;
  /**
   * Image element
   */
  image: HTMLImageElement;
  /**
   * Image data in Base64 ou url
   */
  srcImage: string;
  /**
   * Height
   */
  height: number;
  /**
   * Width
   */
  width: number;
}

export type ILegendRecord = Record<number | string, ILayerLegend[]>;

export interface ILegendSource {
  fetchLegend(): Promise<ILegendRecord>;
}

export interface IExtended extends IInitSource, IQuerySource, ILegendSource { }

export type IGisRequest = IQueryRequest | IIdentifyRequest;

export interface IAbstractRequest<T extends string> {
  olMap: OlMap;
  queryType: T;
  filters?: IPredicate;
  limit?: number;
  returnGeometry?: boolean;
  srId?: string;
}

export interface IQueryRequest extends IAbstractRequest<'query'> {
  geometry?: Geometry;
  geometryProjection?: Projection;
  geometryPrecision?: number;
}

export interface IIdentifyRequest extends IAbstractRequest<'identify'> {
  geometry: Geometry;
  geometryPrecision?: number;
  geometryProjection: Projection;
  identifyTolerance?: number; // in pixel
  layersPrefix?: LayersPrefix;
  outFields?: string;
  returnFieldName?: boolean;
}

export interface IQueryResponse {
  request: IGisRequest;
  featureTypeResponses: IQueryFeatureTypeResponse[];
}

export interface IQueryFeatureTypeResponse {
  type: IFeatureType<any>;
  features: Feature[];
  source: IQuerySource;
}

export interface IAttribute {
  key: string;
  type?: 'Oid' | 'Boolean' | 'Number' | 'String' | 'Date' | 'Geometry' | 'Unknown';
  name?: string;
}

export interface IFeatureType<IDT extends number | string> {
  id: IDT;
  queryId?: IDT;
  hide?: boolean;
  name?: string;
  identifierAttribute?: IAttribute;
  attributes?: IAttribute[];
  readonly initialPredicate?: IPredicate;

  joinWithOr?: boolean;

  predicateWrapper?: FeatureTypePredicateWrapper;
}

export class FeatureTypePredicateWrapper {
  public joinWithOr: boolean;

  private _initialPredicate: IPredicate;
  private _additionalPredicates: Record<string, IPredicate> = {};

  public constructor(initialPredicate?: IPredicate, joinWithOr = true) {
    this._initialPredicate = initialPredicate;
    this.joinWithOr = joinWithOr;
  }

  public addPredicate(predicate: IPredicate): void {
    this._additionalPredicates[predicate.hashCode()] = predicate;
  }

  public removePredicate(hash: string) {
    delete this._additionalPredicates[hash];
  }

  get predicate(): IPredicate {
    return this.filterBuilder.predicate;
  }

  get filterBuilder(): FilterBuilder {
    const predicates: IPredicate[] = Object.values(this._additionalPredicates);
    if (predicates.length > 0) {
      let filterBuilder: FilterBuilder;
      for (const predicate of predicates) {
        if(!filterBuilder) {
          filterBuilder = new FilterBuilder(predicate);
        } else {
          if (this.joinWithOr) {
            filterBuilder.or(predicate);
          } else {
            filterBuilder.and(predicate);
          }
        }
      }
      return filterBuilder.and(this._initialPredicate);
    }
    return new FilterBuilder(this._initialPredicate);
  }

}
export type LayersPrefix = LayersPrefixEnum.ALL | LayersPrefixEnum.TOP | LayersPrefixEnum.VISIBLE;
export enum LayersPrefixEnum {
  ALL = 'all',
  TOP = 'top',
  VISIBLE = 'visible',
}
