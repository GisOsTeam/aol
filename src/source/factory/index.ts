/**
 * @file Automatically generated by barrelsby.
 */

export * from './SourceFacotry';
export * from './WmtsFactory';
