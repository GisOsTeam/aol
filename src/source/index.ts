/**
 * @file Automatically generated by barrelsby.
 */

export * from './ExternalVector';
export * from './Image';
export * from './ImageArcGISRest';
export * from './ImageStatic';
export * from './ImageWms';
export * from './LocalVector';
export * from './QueryArcGISRest';
export * from './TileArcGISRest';
export * from './TileImage';
export * from './TileWms';
export * from './Vector';
export * from './VectorTile';
export * from './Wfs';
export * from './Xyz';
export * from './query/index';
export * from './types/index';
