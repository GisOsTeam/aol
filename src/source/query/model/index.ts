/**
 * @file Automatically generated by barrelsby.
 */

export * from './AgsIdentifyRequest';
export * from './AgsQueryRequest';
