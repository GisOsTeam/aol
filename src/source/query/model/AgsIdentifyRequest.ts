import { IExtended, IFeatureType, IIdentifyRequest, LayersPrefixEnum } from '../../IExtended';
import { transformExtent } from 'ol/proj';
import { getForViewAndSize } from 'ol/extent';
import { FilterBuilder, FilterBuilderTypeEnum } from '../../../filter';
import EsriJSON from 'ol/format/EsriJSON';
import { fromCircle } from 'ol/geom/Polygon';
import Circle from 'ol/geom/Circle';

export interface AgsIdentifyRequestParameters {
  f: string;
  geometry: string;
  geometryType: string;
  sr: string;
  layerDefs: string;
  layers: string;
  tolerance: string;
  mapExtent: string;
  imageDisplay: string;
  returnGeometry: string;
  maxAllowableOffset: string;
  geometryPrecision: string;
  returnZ: string;
  returnM: string;
  returnFieldName: string;
}

export class AgsIdentifyRequest implements AgsIdentifyRequestParameters {
  f: string;
  geometry: string;
  geometryPrecision: string;
  geometryType: string;
  imageDisplay: string;
  layerDefs: string;
  layers: string;
  mapExtent: string;
  maxAllowableOffset: string;
  returnFieldName: string;
  returnGeometry: string;
  returnM: string;
  returnZ: string;
  sr: string;
  tolerance: string;

  private format = new EsriJSON();

  constructor(source: IExtended, types: IFeatureType<number>[], request: IIdentifyRequest) {
    const { olMap, geometryProjection } = request;
    this.sr = '3857';

    if (request.srId) {
      this.sr = request.srId;
    }

    this.returnFieldName = request.returnFieldName ? 'true' : 'false';
    this.returnGeometry = request.returnGeometry ? 'true' : 'false';

    let geometry = request.geometry;
    if (geometry) {
      if (geometry.getType() === 'Circle') {
        geometry = fromCircle(geometry as Circle);
      }
      const geometryStr = this.format.writeGeometry(geometry, {
        featureProjection: geometryProjection,
        dataProjection: 'EPSG:' + this.sr,
      });
      let geometryType = '';
      switch (geometry.getType()) {
        case 'Point':
          geometryType = 'esriGeometryPoint';
          break;
        case 'LineString':
          geometryType = 'esriGeometryPolyline';
          break;
        case 'LinearRing':
          geometryType = 'esriGeometryPolyline';
          break;
        case 'Polygon':
          geometryType = 'esriGeometryPolygon';
          break;
        case 'MultiPoint':
          geometryType = 'esriGeometryMultipoint';
          break;
        case 'MultiLineString':
          geometryType = 'esriGeometryPoint';
          break;
        case 'MultiPolygon':
          geometryType = 'esriGeometryPolygon';
          break;
      }

      this.geometry = geometryStr;
      this.geometryType = geometryType;
      this.geometryPrecision = !!request.geometryPrecision ? `${request.geometryPrecision}` : '10';
    }

    const olView = olMap.getView();

    const { identifyTolerance, layersPrefix = LayersPrefixEnum.ALL } = request;
    let extent = geometry.getExtent();
    if (geometryProjection.getCode() !== 'EPSG:' + this.sr) {
      extent = transformExtent(geometry.getExtent(), geometryProjection, 'EPSG:' + this.sr);
    }
    const mapExtent = getForViewAndSize(
      [0.5 * extent[0] + 0.5 * extent[2], 0.5 * extent[1] + 0.5 * extent[3]],
      olView.getResolution(),
      0,
      [1001, 1001]
    );
    this.mapExtent = mapExtent.join(',');
    this.imageDisplay = '1001,1001';
    const ids = types.map((type) => type.id).join(',');
    this.layers = `${layersPrefix}:${ids}`;
    if (request.filters) {
      const layerDefs = [];
      for (const type of types) {
        const predicate = request.filters[type.id];
        if (predicate) {
          layerDefs.push(`"${type.id}":"${new FilterBuilder(predicate).build(FilterBuilderTypeEnum.SQL)}"`);
        }
      }
      this.layerDefs = `{${layerDefs.join(',')}}`;
    }
    if (!isNaN(Math.round(identifyTolerance))) {
      this.tolerance = `${Math.round(identifyTolerance)}`;
    } else {
      this.tolerance = '4';
    }
    this.f = 'json';
  }

  public getSrId(): string {
    return this.sr;
  }
}
