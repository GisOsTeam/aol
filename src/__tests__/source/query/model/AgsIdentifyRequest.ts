import OlMap from 'ol/Map';
import OlView from 'ol/View';
import { ImageArcGISRest } from '../../../../source';
import { IFeatureType, IIdentifyRequest } from '../../../../source/IExtended';
import Polygon from 'ol/geom/Polygon';
import { get as getProjection } from 'ol/proj';
import { AgsIdentifyRequest } from '../../../../source/query/model/AgsIdentifyRequest';
import Point from 'ol/geom/Point';
import { addProjection } from '../../../../ProjectionInfo';

const states = new ImageArcGISRest({
  url: 'https://sampleserver6.arcgisonline.com/arcgis/rest/services/USA/MapServer',
  types: [{ id: 2, identifierAttribute: { key: 'objectid' } }],
});

const type: IFeatureType<number> = states.get('types')[0];

describe('AgsIdentifyRequest', () => {
  test('default (3857)', () => {
    const request: IIdentifyRequest = {
      olMap: new OlMap({
        view: new OlView({
          center: [0, 0],
          zoom: 1,
        }),
      }),
      geometry: new Polygon([
        [
          [-11580733.168194728, 4880526.966432655],
          [-11580733.168194728, 4964673.98309528],
          [-11496586.151532097, 4964673.983095286],
          [-11496586.151532097, 4880526.966432655],
          [-11580733.168194728, 4880526.966432655],
        ],
      ]),
      geometryProjection: getProjection('EPSG:3857'),
      queryType: 'identify',
      returnGeometry: true,
      returnFieldName: true,
    };
    const agsQueryRequest = new AgsIdentifyRequest(states, [type], request);

    expect<string>(agsQueryRequest.f).toMatchSnapshot();
    expect<string>(agsQueryRequest.geometry).toMatchSnapshot('agsQueryRequest.geometry');
    expect<string>(agsQueryRequest.geometryPrecision).toMatchSnapshot('agsQueryRequest.geometryPrecision');
    expect<string>(agsQueryRequest.geometryType).toMatchSnapshot('agsQueryRequest.geometryType');
    expect<string>(agsQueryRequest.imageDisplay).toMatchSnapshot('agsQueryRequest.imageDisplay');
    expect<string>(agsQueryRequest.layerDefs).toMatchSnapshot('agsQueryRequest.layerDefs');
    expect<string>(agsQueryRequest.layers).toMatchSnapshot('agsQueryRequest.layers');
    expect<string>(agsQueryRequest.mapExtent).toMatchSnapshot('agsQueryRequest.mapExtent');
    expect<string>(agsQueryRequest.maxAllowableOffset).toMatchSnapshot('agsQueryRequest.maxAllowableOffset');
    expect<string>(agsQueryRequest.returnFieldName).toMatchSnapshot('agsQueryRequest.returnFieldName');
    expect<string>(agsQueryRequest.returnGeometry).toMatchSnapshot('agsQueryRequest.returnGeometry');
    expect<string>(agsQueryRequest.returnM).toMatchSnapshot('agsQueryRequest.returnM');
    expect<string>(agsQueryRequest.returnZ).toMatchSnapshot('agsQueryRequest.returnZ');
    expect<string>(agsQueryRequest.sr).toMatchSnapshot('agsQueryRequest.sr');
    expect<string>(agsQueryRequest.tolerance).toMatchSnapshot('agsQueryRequest.tolerance');
  });

  test('2154', () => {
    // addProjection(
    //   'EPSG:2154',
    //   '+proj=lcc +lat_1=49 +lat_2=44 +lat_0=46.5 +lon_0=3 +x_0=700000 +y_0=6600000 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 +units=m +no_defs'
    // );
    // const request: IIdentifyRequest = {
    //   olMap: new OlMap({
    //     view: new OlView({
    //       center: [0, 0],
    //       zoom: 1,
    //     }),
    //   }),
    //   geometry: new Point([827008.83, 6506549.89]),
    //   geometryProjection: getProjection('EPSG:2154'),
    //   queryType: 'identify',
    //   srId: '2154',
    // };
    // const agsQueryRequest = new AgsIdentifyRequest(states, [type], request);
    // expect<string>(agsQueryRequest.f).toMatchSnapshot();
    // expect<string>(agsQueryRequest.geometry).toMatchSnapshot('agsQueryRequest.geometry');
    // expect<string>(agsQueryRequest.geometryPrecision).toMatchSnapshot('agsQueryRequest.geometryPrecision');
    // expect<string>(agsQueryRequest.geometryType).toMatchSnapshot('agsQueryRequest.geometryType');
    // expect<string>(agsQueryRequest.imageDisplay).toMatchSnapshot('agsQueryRequest.imageDisplay');
    // expect<string>(agsQueryRequest.layerDefs).toMatchSnapshot('agsQueryRequest.layerDefs');
    // expect<string>(agsQueryRequest.layers).toMatchSnapshot('agsQueryRequest.layers');
    // expect<string>(agsQueryRequest.mapExtent).toMatchSnapshot('agsQueryRequest.mapExtent');
    // expect<string>(agsQueryRequest.maxAllowableOffset).toMatchSnapshot('agsQueryRequest.maxAllowableOffset');
    // expect<string>(agsQueryRequest.returnFieldName).toMatchSnapshot('agsQueryRequest.returnFieldName');
    // expect<string>(agsQueryRequest.returnGeometry).toMatchSnapshot('agsQueryRequest.returnGeometry');
    // expect<string>(agsQueryRequest.returnM).toMatchSnapshot('agsQueryRequest.returnM');
    // expect<string>(agsQueryRequest.returnZ).toMatchSnapshot('agsQueryRequest.returnZ');
    // expect<string>(agsQueryRequest.sr).toMatchSnapshot('agsQueryRequest.sr');
    // expect<string>(agsQueryRequest.tolerance).toMatchSnapshot('agsQueryRequest.tolerance');
  });

  test('3296', () => {
    // addProjection(
    //   'EPSG:3296',
    //   '+proj=utm +zone=5 +south +ellps=GRS80 +towgs84=0.072,-0.507,-0.245,-0.0183,0.0003,-0.007,-0.0093 +units=m +no_defs'
    // );
    // const request: IIdentifyRequest = {
    //   olMap: new OlMap({
    //     view: new OlView({
    //       center: [0, 0],
    //       zoom: 1,
    //     }),
    //   }),
    //   geometry: new Point([632635.3571388097, 8180019.336045767]),
    //   geometryProjection: getProjection('EPSG:3296'),
    //   queryType: 'identify',
    //   srId: '3296',
    //   geometryPrecision: 20,
    // };
    // const agsQueryRequest = new AgsIdentifyRequest(states, [type], request);
    // expect<string>(agsQueryRequest.f).toMatchSnapshot();
    // expect<string>(agsQueryRequest.geometry).toMatchSnapshot('agsQueryRequest.geometry');
    // expect<string>(agsQueryRequest.geometryPrecision).toMatchSnapshot('agsQueryRequest.geometryPrecision');
    // expect<string>(agsQueryRequest.geometryType).toMatchSnapshot('agsQueryRequest.geometryType');
    // expect<string>(agsQueryRequest.imageDisplay).toMatchSnapshot('agsQueryRequest.imageDisplay');
    // expect<string>(agsQueryRequest.layerDefs).toMatchSnapshot('agsQueryRequest.layerDefs');
    // expect<string>(agsQueryRequest.layers).toMatchSnapshot('agsQueryRequest.layers');
    // expect<string>(agsQueryRequest.mapExtent).toMatchSnapshot('agsQueryRequest.mapExtent');
    // expect<string>(agsQueryRequest.maxAllowableOffset).toMatchSnapshot('agsQueryRequest.maxAllowableOffset');
    // expect<string>(agsQueryRequest.returnFieldName).toMatchSnapshot('agsQueryRequest.returnFieldName');
    // expect<string>(agsQueryRequest.returnGeometry).toMatchSnapshot('agsQueryRequest.returnGeometry');
    // expect<string>(agsQueryRequest.returnM).toMatchSnapshot('agsQueryRequest.returnM');
    // expect<string>(agsQueryRequest.returnZ).toMatchSnapshot('agsQueryRequest.returnZ');
    // expect<string>(agsQueryRequest.sr).toMatchSnapshot('agsQueryRequest.sr');
    // expect<string>(agsQueryRequest.tolerance).toMatchSnapshot('agsQueryRequest.tolerance');
  });
});
