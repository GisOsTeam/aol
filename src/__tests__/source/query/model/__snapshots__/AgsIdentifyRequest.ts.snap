// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AgsIdentifyRequest 2154 1`] = `"json"`;

exports[`AgsIdentifyRequest 2154: agsQueryRequest.geometry 1`] = `"{\\"x\\":827008.83,\\"y\\":6506549.89}"`;

exports[`AgsIdentifyRequest 2154: agsQueryRequest.geometryPrecision 1`] = `"10"`;

exports[`AgsIdentifyRequest 2154: agsQueryRequest.geometryType 1`] = `"esriGeometryPoint"`;

exports[`AgsIdentifyRequest 2154: agsQueryRequest.imageDisplay 1`] = `"1001,1001"`;

exports[`AgsIdentifyRequest 2154: agsQueryRequest.layerDefs 1`] = `undefined`;

exports[`AgsIdentifyRequest 2154: agsQueryRequest.layers 1`] = `"all:2"`;

exports[`AgsIdentifyRequest 2154: agsQueryRequest.mapExtent 1`] = `"-38347885.41049226,-32668344.350492254,40001903.07049225,45681444.130492255"`;

exports[`AgsIdentifyRequest 2154: agsQueryRequest.maxAllowableOffset 1`] = `undefined`;

exports[`AgsIdentifyRequest 2154: agsQueryRequest.returnFieldName 1`] = `"true"`;

exports[`AgsIdentifyRequest 2154: agsQueryRequest.returnGeometry 1`] = `"true"`;

exports[`AgsIdentifyRequest 2154: agsQueryRequest.returnM 1`] = `undefined`;

exports[`AgsIdentifyRequest 2154: agsQueryRequest.returnZ 1`] = `undefined`;

exports[`AgsIdentifyRequest 2154: agsQueryRequest.sr 1`] = `"2154"`;

exports[`AgsIdentifyRequest 2154: agsQueryRequest.tolerance 1`] = `"4"`;

exports[`AgsIdentifyRequest 3296 1`] = `"json"`;

exports[`AgsIdentifyRequest 3296: agsQueryRequest.geometry 1`] = `"{\\"x\\":632635.3571388097,\\"y\\":8180019.336045767}"`;

exports[`AgsIdentifyRequest 3296: agsQueryRequest.geometryPrecision 1`] = `"20"`;

exports[`AgsIdentifyRequest 3296: agsQueryRequest.geometryType 1`] = `"esriGeometryPoint"`;

exports[`AgsIdentifyRequest 3296: agsQueryRequest.imageDisplay 1`] = `"1001,1001"`;

exports[`AgsIdentifyRequest 3296: agsQueryRequest.layerDefs 1`] = `undefined`;

exports[`AgsIdentifyRequest 3296: agsQueryRequest.layers 1`] = `"all:2"`;

exports[`AgsIdentifyRequest 3296: agsQueryRequest.mapExtent 1`] = `"-38542258.88335344,-30994874.904446486,39807529.59763107,47354913.57653802"`;

exports[`AgsIdentifyRequest 3296: agsQueryRequest.maxAllowableOffset 1`] = `undefined`;

exports[`AgsIdentifyRequest 3296: agsQueryRequest.returnFieldName 1`] = `"true"`;

exports[`AgsIdentifyRequest 3296: agsQueryRequest.returnGeometry 1`] = `"true"`;

exports[`AgsIdentifyRequest 3296: agsQueryRequest.returnM 1`] = `undefined`;

exports[`AgsIdentifyRequest 3296: agsQueryRequest.returnZ 1`] = `undefined`;

exports[`AgsIdentifyRequest 3296: agsQueryRequest.sr 1`] = `"3296"`;

exports[`AgsIdentifyRequest 3296: agsQueryRequest.tolerance 1`] = `"4"`;

exports[`AgsIdentifyRequest default (3857) 1`] = `"json"`;

exports[`AgsIdentifyRequest default (3857): agsQueryRequest.geometry 1`] = `"{\\"hasZ\\":false,\\"hasM\\":false,\\"rings\\":[[[-11580733.168194728,4880526.966432655],[-11580733.168194728,4964673.98309528],[-11496586.151532097,4964673.983095286],[-11496586.151532097,4880526.966432655],[-11580733.168194728,4880526.966432655]]]}"`;

exports[`AgsIdentifyRequest default (3857): agsQueryRequest.geometryPrecision 1`] = `"10"`;

exports[`AgsIdentifyRequest default (3857): agsQueryRequest.geometryType 1`] = `"esriGeometryPolygon"`;

exports[`AgsIdentifyRequest default (3857): agsQueryRequest.imageDisplay 1`] = `"1001,1001"`;

exports[`AgsIdentifyRequest default (3857): agsQueryRequest.layerDefs 1`] = `"{}"`;

exports[`AgsIdentifyRequest default (3857): agsQueryRequest.layers 1`] = `"all:2"`;

exports[`AgsIdentifyRequest default (3857): agsQueryRequest.mapExtent 1`] = `"-50713553.90035567,-34252293.76572828,27636234.580628842,44097494.71525623"`;

exports[`AgsIdentifyRequest default (3857): agsQueryRequest.maxAllowableOffset 1`] = `undefined`;

exports[`AgsIdentifyRequest default (3857): agsQueryRequest.returnFieldName 1`] = `"true"`;

exports[`AgsIdentifyRequest default (3857): agsQueryRequest.returnGeometry 1`] = `"true"`;

exports[`AgsIdentifyRequest default (3857): agsQueryRequest.returnM 1`] = `undefined`;

exports[`AgsIdentifyRequest default (3857): agsQueryRequest.returnZ 1`] = `undefined`;

exports[`AgsIdentifyRequest default (3857): agsQueryRequest.sr 1`] = `"3857"`;

exports[`AgsIdentifyRequest default (3857): agsQueryRequest.tolerance 1`] = `"4"`;
